File: src/main/java/com/atsuishio/superbwarfare/perk/functional/CastNoShadows.java
package com.atsuishio.superbwarfare.perk.functional;

import com.atsuishio.superbwarfare.capability.player.PlayerVariable;
import com.atsuishio.superbwarfare.data.gun.GunData;
import com.atsuishio.superbwarfare.data.gun.GunProp;
import com.atsuishio.superbwarfare.perk.Perk;
import com.atsuishio.superbwarfare.perk.PerkInstance;
import com.atsuishio.superbwarfare.tools.InventoryTool;
import net.minecraft.world.damagesource.DamageSource;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.entity.projectile.Projectile;

public class CastNoShadows extends Perk {
    public CastNoShadows() {
        super("cast_no_shadows", Perk.Type.FUNCTIONAL);
    }

    @Override
    public void onMeleeAttack(GunData data, PerkInstance instance, Entity target, DamageSource source) {
        super.onMeleeAttack(data, instance, target, source);

        Player attacker = null;
        if (source.getEntity() instanceof Player player) {
            attacker = player;
        }
        if (source.getDirectEntity() instanceof Projectile p && p.getOwner() instanceof Player player) {
            attacker = player;
        }

        if (attacker == null) return;

        float rate = 0.2f + (instance.level() - 1) * 0.03f;

        Player finalAttacker = attacker;
        PlayerVariable.modify(attacker, cap -> {
            int mag = data.get(GunProp.MAGAZINE);
            int ammo = data.ammo.get();
            int ammoReload = (int) Math.min(mag, mag * rate);
            int ammoNeed = Math.min(mag - ammo, ammoReload);

            boolean flag = finalAttacker.isCreative() || InventoryTool.hasCreativeAmmoBox(finalAttacker);

            int ammoFinal = Math.min(data.countBackupAmmo(finalAttacker), ammoNeed);
            if (flag) {
                ammoFinal = ammoNeed;
            } else {
                data.consumeBackupAmmo(finalAttacker, ammoFinal);
            }
            data.ammo.set(Math.min(mag, ammo + ammoFinal));
        });
    }
}

File: src/main/resources/data/superbwarfare/sbw/perks/cast_no_shadows.js
function onMeleeAttack(tag, level, gunData, target, source) {
    var attacker = source.getSourceEntity()
    if (!attacker.isPlayer()) return

    var rate = 0.2 + (level - 1) * 0.03
    attacker.heal(attacker.getMaxHealth() * rate / 2)

    var mag = gunData.getMagazine()
    var ammo = gunData.getAmmo()
    var ammoReload = Math.min(mag, Math.floor(mag * rate))
    var ammoNeed = Math.min(mag - ammo, ammoReload)

    var flag = attacker.isCreative() || attacker.hasCreativeAmmoBox()
    var ammoFinal = Math.min(gunData.countBackupAmmo(attacker), ammoNeed)

    if (flag) {
        ammoFinal = ammoNeed
    } else {
        gunData.consumeBackupAmmo(attacker, ammoFinal)
    }
    gunData.setAmmo(Math.min(mag, ammo + ammoFinal))
}
